### PR TITLE
surrealist: init at 1.11.5

### PR DESCRIPTION
**Pull request by MDM23**
## Description of changes

This provides the [Surrealist](https://surrealist.starlane.studio) application, as requested in [#255193](https://github.com/NixOS/nixpkgs/issues/255193).

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/top-level/all-packages.nix
Comment (AndersonTorres) [inline pkgs/top-level/all-packages.nix:20165]: move the default.nix file to pkgs/by-name/su/surrealist/package.nix

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/top-level/all-packages.nix
Comment (MDM23) [reply pkgs/top-level/all-packages.nix]: Moved it :+1:

File: pkgs/by-name/su/surrealist/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/su/surrealist/package.nix:26]:
```suggestion
  unpackPhase = ''
    runHook preUnpack
    ${dpkg}/bin/dpkg -x $src ./
    runHook postUnpack
  '';
```

Or even

```suggestion
  unpackCmd = ''
    ${dpkg}/bin/dpkg -x $src ./
  '';
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Review by AndersonTorres — APPROVED** [review]

**Review by AndersonTorres — APPROVED** [review]
Looks nice.

Compiling from source is too hard?

File: pkgs/by-name/su/surrealist/package.nix
Comment (h7x4) [inline pkgs/by-name/su/surrealist/package.nix:15]: It doesn't seem like this uses the C toolchain, since we're downloading precompiled. Maybe this would work as well?

```suggestion
, stdenvNoCC
, webkitgtk
}:

stdenvNoCC.mkDerivation rec {
```

File: pkgs/by-name/su/surrealist/package.nix
Comment (h7x4) [inline pkgs/by-name/su/surrealist/package.nix:50]: nit: no need for write permissions, since nix store is read only.

```suggestion
    install -Dm555 usr/bin/surrealist -t $out/bin
    cp -r usr/share $out
    chmod -R 444 $out/share
```

File: pkgs/by-name/su/surrealist/package.nix
Comment (AndersonTorres) [reply pkgs/by-name/su/surrealist/package.nix:15]: Not necessarily. Sometimes the binary is autopatchelf'ed.

**Comment by MDM23** [comment]
> Compiling from source is too hard?

Not trivial at least. They use [pnpm](https://pnpm.io) to build the UI and there is no regular lockfile that could be used with npm. I'll give it a try and check how some other packages handled this.

Some examples:
- [geph](https://github.com/NixOS/nixpkgs/blob/3de3f7d87956a2dba7506b65e1c6709e3d6cb377/pkgs/applications/networking/geph/default.nix)
- [pot](https://github.com/NixOS/nixpkgs/blob/3de3f7d87956a2dba7506b65e1c6709e3d6cb377/pkgs/applications/misc/pot/default.nix)
- [vesktop](https://github.com/NixOS/nixpkgs/blob/e97b3e4186bcadf0ef1b6be22b8558eab1cdeb5d/pkgs/by-name/ve/vesktop/package.nix)

**Comment by MDM23** [comment]
I rewrote it to be built from source now and have pushed an update.

**Comment by silvergasp** [comment]
Not a maintainer, so I can't merge this. But I downloaded this PR and have been using this surrealist package locally for a couple of weeks without any issues. Hopefully that helps with making a decision to merge/or not :)

**Comment by nyabinary** [comment]
Any updates  on this :3

**Comment by MDM23** [comment]
I'm currently working on packaging version 1.11.2. Surrealist embeds a pre-compiled wasm library (its source is located under [src-embed](https://github.com/StarlaneStudios/Surrealist/tree/6275e6b2a95c01c895a187aa850cb0810e4a8529/src-embed)). Those artifacts used to be shipped in the repository but they have been removed in the new version. It is now done as an extra build step.

**Comment by nyabinary** [comment]
> I'm currently working on packaging version 1.11.2. Surrealist embeds a pre-compiled wasm library (its source is located under [src-embed](https://github.com/StarlaneStudios/Surrealist/tree/6275e6b2a95c01c895a187aa850cb0810e4a8529/src-embed)). Those artifacts used to be shipped in the repository but they have been removed in the new version. It is now done as an extra build step.

any update on this?

**Comment by nyabinary** [comment]
> Bumped to v1.11.3

nice :3

File: pkgs/by-name/su/surrealist/package.nix
Comment (drupol) [inline pkgs/by-name/su/surrealist/package.nix:53]: If you are overriding `configurePhase`, `buildPhase`, `checkPhase`, `installPhase` or any other phase, you should not forget about explicitly running pre and post-phase hooks like [their](https://github.com/NixOS/nixpkgs/blob/d71a03ad695407dd482ead32d3eddff50092a1c3/pkgs/stdenv/generic/setup.sh#L953) [original](https://github.com/NixOS/nixpkgs/blob/d71a03ad695407dd482ead32d3eddff50092a1c3/pkgs/stdenv/generic/setup.sh#L1008) [definitions](https://github.com/NixOS/nixpkgs/blob/d71a03ad695407dd482ead32d3eddff50092a1c3/pkgs/stdenv/generic/setup.sh#L1037) [do](https://github.com/NixOS/nixpkgs/blob/d71a03ad695407dd482ead32d3eddff50092a1c3/pkgs/stdenv/generic/setup.sh#L1078).

It is generally expected that an appropriate pre-phase hook (e.g. `preBuild`) hook will run at the beginning of a phase (e.g. `buildPhase`) and post-phase hook (e.g. `postBuild`) will run at the end. Hooks are normally ran as a part of a phase so if you override a phase as a whole, you will need to add `runHook hookName` calls (e.g. `runHook preBuild`) manually.

Having phases run pre/post-phase hooks is important because many setup hooks insert their own code into them – omitting a hook might therefore prevent some setup hooks required for proper functionality of a package from running. Additionally, hooks are often inserted by developers into the package expression and by users when overriding a package using `overrideAttrs`. Not running them can thus cause confusion why their code is not executed.

## Examples

### Before

```nix
  installPhase = ''
    your commands
  '';
```

### After

```nix
  installPhase = ''
    runHook preInstall

    your commands

    runHook postInstall
  '';
```

Source: https://github.com/jtojnar/nixpkgs-hammering/blob/main/explanations/missing-phase-hooks.md

**Review by drupol — CHANGES_REQUESTED** [review]
Hi !

Thanks for your contribution.

I've left some comments, let me know if this is clear enough.

**Review by drupol — CHANGES_REQUESTED** [review]
There's a build issue on platform `aarch64-linux`, do you think you could have a look?

**Review by drupol — APPROVED** [review]

**Pull request merged**
